src/dag.rs

```diff
@@ -30,43 +174,414 @@ impl<Types: RepoTypes> IpldDag<Types> {
+    async fn resolve_hamt(
+        &self,
+        mut lookup: ShardedLookup<'_>,
+        cache: &mut Option<Cache>,
+    ) -> Result<Cid, Error> {
+        use MaybeResolved::*;
+
+        loop {
+            let (next, _) = lookup.pending_links();
+
+            let block = self.ipfs.repo.get_block(next).await?;
+
+            match lookup.continue_walk(block.data(), cache)? {
+                NeedToLoadMore(next) => lookup = next,
+                Found(cid) => return Ok(cid),
+                NotFound => return Err(anyhow::anyhow!("key not found: ???")),
+            }
         }
-        Ok(ipld)
-    }
-}
-
-fn try_resolve(ipld: Ipld, segment: &str) -> Result<Ipld, Error> {
-    match ipld {
-        Ipld::Map(mut map) => map
-            .remove(segment)
-            .ok_or_else(|| anyhow::anyhow!("no such segment: {:?}", segment)),
-        Ipld::List(mut vec) => match segment.parse::<usize>() {
-            Ok(index) if index < vec.len() => Ok(vec.swap_remove(index)),
-            Ok(_) => Err(anyhow::anyhow!(
-                "index out of range 0..{}: {:?}",
-                vec.len(),
-                segment
-            )),
-            Err(_) => Err(anyhow::anyhow!("invalid list index: {:?}", segment)),
-        },
-        link @ Ipld::Link(_) if segment == "." => Ok(link),
-        other => Err(anyhow::anyhow!(
-            "cannot resolve {:?} through: {:?}",
+    }
+}
+
+/// `IpfsPath`'s cid based variant can be resolved to the block, projections represented by this
+/// type.
+///
+/// Values can be converted to Ipld using `Ipld::try_from`.
+#[derive(Debug, PartialEq)]
+pub enum ResolvedNode {
+    /// Block which was loaded at the end of the path.
+    Block(Block),
+    /// Path ended in `Data` at a dag-pb node. This is usually not interesting and should be
+    /// treated as "Not found" error since dag-pb node did not have a *link* called `Data`. Variant
+    /// exists as there are interface-ipfs-http tests which require this behaviour.
+    DagPbData(Cid, NodeData<Box<[u8]>>),
+    /// Path ended on an !dag-pb document which was projected.
+    Projection(Cid, Ipld),
+    /// Local resolving ended in at link
+    Link(Cid, Cid),
+}
+
+impl ResolvedNode {
+    /// Returns the cid of the **source** document. Source or destination matters only the case of link
+    /// to another document.
+    pub fn source(&self) -> &Cid {
+        match self {
+            ResolvedNode::Block(Block { cid, .. })
+            | ResolvedNode::DagPbData(cid, ..)
+            | ResolvedNode::Projection(cid, ..)
+            | ResolvedNode::Link(cid, ..) => cid,
+        }
+    }
+
+    /// Returns the cid of the linked document when resolved to a link, otherwise `None`.
+    pub fn destination(&self) -> Option<&Cid> {
+        match self {
+            ResolvedNode::Link(_, cid) => Some(cid),
+            _ => None,
+        }
+    }
+
+    /// Returns the destination or the source link.
```

Comment:
Also, I was a bit surprised that this wasn't "the destination **of** the source link" - perhaps it's just me being confused a bit with the naming here, but being able to return _either_ a destination _or_ a source X seems weird at first glance.

Reply:
The other thread: https://github.com/rs-ipfs/rust-ipfs/pull/332#discussion_r476348063